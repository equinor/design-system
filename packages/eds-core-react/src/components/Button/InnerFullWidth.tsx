import { forwardRef, Children as ReactChildren, HTMLAttributes } from 'react'
import styled from 'styled-components'

const FullWidthCenterContent = styled.span`
  text-align: center;
  flex: 1;
`

const FullWidthInner = styled.span`
  height: 100%;
  display: flex;
  align-items: center;

  > :is(svg, img) {
    margin-top: var(--eds_button__margin_y, inherit);
    margin-bottom: var(--eds_button__margin_y, inherit);
  }

  > :is(svg, img):first-child {
    margin-right: var(--eds_button__fullwidth__icon__margin_x, 8px);
  }

  > :is(svg, img):last-child {
    margin-left: var(--eds_button__fullwidth__icon__margin_x, 8px);
  }

  > :is(svg, img):only-child {
    margin-left: auto;
    margin-right: auto;
  }

  > span:first-child {
    margin-left: var(--eds_button__fullwidth__margin_x, 32px);
  }

  > span:last-child {
    margin-right: var(--eds_button__fullwidth__margin_x, 32px);
  }

  > span:only-child {
    margin-right: 0;
    margin-left: 0;
  }
`

export const InnerFullWidth = forwardRef<
  HTMLSpanElement,
  HTMLAttributes<HTMLSpanElement>
>(function InnerFullWidth({ children }, ref) {
  // We need everything in elements for proper flexing ðŸ’ª
  const updatedChildren = ReactChildren.map(children, (child) =>
    typeof child !== 'object' ? (
      <FullWidthCenterContent>{child}</FullWidthCenterContent>
    ) : (
      child
    ),
  )

  return <FullWidthInner ref={ref}>{updatedChildren}</FullWidthInner>
})
