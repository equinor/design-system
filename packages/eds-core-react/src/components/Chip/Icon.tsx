import styled, { css } from 'styled-components'
import { close } from '@equinor/eds-icons'
import { Icon as Icon_ } from '../Icon'
import * as tokens from './Chip.tokens'
import { bordersTemplate } from '@equinor/eds-utils'

Icon_.add({ close })

const { enabled, error } = tokens

type IconProps = {
  $variant: 'active' | 'error' | 'default'
  $disabled: boolean
}

export const Icon = styled(Icon_)<IconProps>`
  cursor: pointer;
  padding: 1px;
  ${bordersTemplate(enabled.entities.icon.border)}
  &:hover {
    ${({ $variant }) => {
      switch ($variant) {
        case 'error':
          return css`
            background: ${error.entities.icon.background};
            svg {
              fill: ${error.entities.icon.typography.color};
            }
          `
        default:
          return css`
            background: ${enabled.entities.icon.states.hover.background};
            svg {
              fill: ${enabled.states.hover.typography.color};
            }
          `
      }
    }};
  }

  ${({ $disabled }) =>
    $disabled &&
    css`
      cursor: not-allowed;
      &:hover {
        background: transparent;
      }
    `}
`
