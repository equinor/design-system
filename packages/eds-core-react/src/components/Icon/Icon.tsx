import { forwardRef, Ref, SVGProps } from 'react'
import type { IconData } from '@equinor/eds-icons'
import styled from 'styled-components'
import type { Name } from './Icon.types'
import { get } from './library'

type StyledProps = {
  height: number
  $height?: number
  width: number
  fill: string
  size?: number
  rotation?: number
}

type SimpleSVGProps = {
  name: string
  viewBox: string
  rotation?: number
  title?: string
  role?: string
  'aria-hidden'?: boolean | 'true' | 'false'
  'aria-labelledby'?: string
}

const transform = ({ rotation }: SimpleSVGProps): string =>
  rotation ? `transform: rotate(${rotation}deg)` : ''

const StyledSvg = styled.svg.attrs<StyledProps>(({ height, width, fill }) => ({
  name: null,
  xmlns: 'http://www.w3.org/2000/svg',
  height: `${height}px`,
  width: `${width}px`,
  fill,
}))`
  ${transform}
`

const StyledPath = styled.path.attrs<StyledProps>(({ $height, size }) => ({
  size: null,
  fillRule: 'evenodd',
  clipRule: 'evenodd',
  transform: size / $height !== 1 ? `scale(${size / $height})` : null,
}))``

const customIcon = (icon: IconData) => ({
  icon,
  count: Math.floor(Math.random() * 1000),
})

const findIcon = (name: string, data: IconData, size: number) => {
  // eslint-disable-next-line prefer-const
  let { icon, count } = data ? customIcon(data) : get(name)

  if (size < 24) {
    // fallback to normal icon if small is not made yet
    icon = icon.sizes?.small || icon
  }

  return { icon, count }
}

export type IconProps = {
  /** Title for icon when used semantically */
  title?: string
  /** Color */
  color?: string
  /** Size */
  size?: 16 | 18 | 24 | 32 | 40 | 48
  /** Rotation */
  rotation?: 0 | 90 | 180 | 270
  /** Name */
  name?: Name
  /** Manually specify which icon data to use */
  data?: IconData
  /** @ignore */
  ref?: Ref<SVGSVGElement>
} & SVGProps<SVGSVGElement>

export const Icon = forwardRef<SVGSVGElement, IconProps>(function Icon(
  { size, color = 'currentColor', name, rotation, title, data, ...rest },
  ref,
) {
  // eslint-disable-next-line prefer-const
  const { icon, count } = findIcon(name, data, size)

  if (typeof icon === 'undefined') {
    throw Error(
      `Icon "${name}" not found. Have you added it using Icon.add() or using data props?`,
    )
  }

  const height = size ? size : parseInt(icon.width)
  const width = size ? size : parseInt(icon.height)

  let svgProps: SimpleSVGProps & StyledProps = {
    height,
    width,
    fill: color,
    viewBox: `0 0 ${width} ${height}`,
    rotation,
    name,
    'aria-hidden': true,
  }

  const pathProps = {
    d: icon.svgPathData,
    $height: icon.height ? icon.height : size,
    size: size || icon.height,
  }

  // Accessibility
  let titleId = ''

  if (title) {
    titleId = `${icon.prefix}-${icon.name}-${count}`
    svgProps = {
      ...svgProps,
      title,
      role: 'img',
      'aria-hidden': undefined,
      'aria-labelledby': titleId,
    }
  }

  return (
    <StyledSvg {...svgProps} {...rest} ref={ref}>
      {title && <title id={titleId}>{title}</title>}
      <StyledPath data-testid="eds-icon-path" {...pathProps} />
    </StyledSvg>
  )
})

// Icon.displayName = 'EdsIcon'
