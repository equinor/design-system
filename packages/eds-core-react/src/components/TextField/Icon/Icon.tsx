import { HTMLAttributes, forwardRef } from 'react'
import styled, { css } from 'styled-components'
import { useTextField } from '../TextField.context'
import { input as tokens } from './Icon.tokens'
import type { Variants, ColorStateProps } from '../types'

type StyledIconProps = {
  colors: ColorStateProps
  isDisabled?: boolean
  isFocused?: boolean
  size?: 16 | 24
}

const StyledIcon = styled.div<StyledIconProps>(
  ({ colors, isDisabled, isFocused, size }) => {
    const { focusColor, color, disabledColor } = colors
    let fill = color

    if (isDisabled) {
      fill = disabledColor
    }
    if (isFocused) {
      fill = focusColor
    }

    return css`
      &,
      svg {
        fill: ${fill};
        width: ${size}px;
        height: ${size}px;
      }
    `
  },
)

type TextfieldIconProps = {
  /** isDisabled */
  isDisabled?: boolean
  /** Variant */
  variant?: Variants
  /** Colors */
  colors?: ColorStateProps
  /** Size */
  size?: 16 | 24
} & HTMLAttributes<HTMLDivElement>

const InputIcon = forwardRef<HTMLDivElement, TextfieldIconProps>(
  function InputIcon(
    {
      size = 24,
      variant = 'default',
      isDisabled = false,
      colors = {
        color: tokens[variant].color,
        disabledColor: tokens[variant].disabledColor,
        focusColor: tokens[variant].focusColor,
      },
      children,
      ...other
    },
    ref,
  ) {
    const { isFocused } = useTextField()

    const iconProps = {
      isDisabled,
      colors,
      isFocused,
      size,
    }

    return (
      <StyledIcon ref={ref} {...iconProps} {...other}>
        {children}
      </StyledIcon>
    )
  },
)

export { InputIcon as Icon }
