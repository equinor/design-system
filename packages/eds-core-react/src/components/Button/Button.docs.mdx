
import { Story, PropsTable, Links } from './../../../.storybook/components'
import { Button  } from '.'


# Button 
Buttons allow users to take action with a single click or tap

<Links 
ariaUrl="https://www.w3.org/WAI/ARIA/apg/patterns/button/" 
figmaUrl="https://www.figma.com/file/0bGXR2sCwMVSDNyyzu5BXrO5/UI%E2%80%94User-Interface?node-id=17394%3A47297" 
documentationUrl="https://eds.equinor.com/0b0c666ab/p/83185a-button/b/86826a"
sourceUrl="https://github.com/equinor/design-system/tree/develop/packages/eds-core-react/src/components/Button"
npmUrl="https://www.npmjs.com/package/@equinor/eds-core-react/"
/>

<Story id="inputs-button--introduction" />
<PropsTable />


## Usage

```tsx
import { Button } from '@equinor/eds-core-react'
<Button>Click me!</Button>
```
## Basic

The `Button` come with four variants; `contained`, `outlined`, `ghost` and `ghost_icon`.
<Story id="inputs-button--basic" />

## Icon button

The `ghost_icon` variant is meant for standalone icon buttons using the `<Icon>` component.


<Story id="inputs-button--ghost-icon" />


```tsx
<Button variant="ghost_icon">
  <Icon name="save" title="save action"></Icon>
</Button>
```

## Color
<Story id="inputs-button--color"/>

## Compact

Compact `Button` using `EdsProvider`
<Story id="inputs-button--compact" />

## File upload

Example of how to do file upload button. Please note this demo only works in Storybook Canvas (isolated example only).

<Story id="inputs-button--file-upload" />

```tsx

<label htmlFor="file-upload">
  <input type="file" id="file-upload" style={{ display: 'none' }} multiple />
  <Button as="span">Upload file</Button>
</label>
```

## Progress button

Use the `Progress` component if you need progress indication on a button. Make sure the 
<Story id="inputs-button--progress-button" />


## All

Example All button combinations. 

<Story id="inputs-button--all" />

## Full width

Examples of buttons
<Story id="inputs-button--full-width" />
