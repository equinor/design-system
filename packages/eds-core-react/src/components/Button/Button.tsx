import { forwardRef, ButtonHTMLAttributes } from 'react'
import styled, { css, ThemeProvider } from 'styled-components'
import { token as buttonToken } from './tokens'
import { ButtonTokenSet, ButtonToken } from './Button.types'
import {
  typographyTemplate,
  bordersTemplate,
  outlineTemplate,
  spacingsTemplate,
  useToken,
  OverridableComponent,
} from '@equinor/eds-utils'
import { InnerFullWidth } from './InnerFullWidth'
import { useEds } from '../EdsProvider'

type Colors = 'primary' | 'secondary' | 'danger'
type Variants =
  | 'contained'
  | 'contained_icon'
  | 'outlined'
  | 'ghost'
  | 'ghost_icon'

const getVariant = (
  tokenSet: ButtonTokenSet,
  variant: Variants,
): ButtonToken => {
  switch (variant) {
    case 'ghost':
      return tokenSet.ghost
    case 'ghost_icon':
      return tokenSet.ghost_icon
    case 'outlined':
      return tokenSet.outlined
    case 'contained_icon':
      return tokenSet.contained_icon
    case 'contained':
    default:
      return tokenSet.contained
  }
}

const getToken = (variant: Variants, color: Colors): ButtonToken => {
  switch (color) {
    case 'danger':
      return getVariant(buttonToken.danger, variant)
    case 'secondary':
      return getVariant(buttonToken.secondary, variant)
    case 'primary':
    default:
      return getVariant(buttonToken.primary, variant)
  }
}

const Inner = styled.span`
  display: grid;
  grid-gap: var(--eds_button__gap, 8px);
  grid-auto-flow: column;
  align-items: center;
  height: 100%;
  justify-content: center;

  & > :is(svg, img) {
    margin-top: var(--eds_button__icon__margin_y, 0);
    margin-bottom: var(--eds_button__icon__margin_y, 0);
  }
`

const ButtonBase = styled.button(({ theme }: { theme: ButtonToken }) => {
  const { states, clickbound } = theme
  const { focus, hover, disabled } = states

  return css`
    box-sizing: border-box;
    margin: 0;
    padding: 0;
    text-decoration: none;
    position: relative;
    cursor: pointer;
    display: inline-block;
    background: ${theme.background};
    height: ${theme.height};
    width: ${theme.width};

    svg {
      justify-self: center;
    }

    ${spacingsTemplate(theme.spacings)}
    ${bordersTemplate(theme.border)}
    ${typographyTemplate(theme.typography)}

    &::before {
      position: absolute;
      top: 0;
      left: 0;
      width: auto;
      min-height: auto;
      content: '';
    }

    &::after {
      position: absolute;
      top: -${clickbound?.offset?.top};
      left: -${clickbound?.offset?.left};
      width: ${clickbound?.width};
      height: ${clickbound?.height};
      content: '';
    }

    @media (hover: hover) and (pointer: fine) {
      &:hover {
        background: ${hover.background};
        color: ${hover.typography?.color};
        ${bordersTemplate(hover?.border)};
      }
    }

    &:focus {
      outline: none;
    }

    &[data-focus-visible-added]:focus {
      ${outlineTemplate(focus.outline)}
    }
    &:focus-visible {
      ${outlineTemplate(focus.outline)}
    }
    /* Get rid of ff focus border for buttons */
    &::-moz-focus-inner {
      border: 0;
    }

    &:disabled,
    &[aria-disabled='true'] {
      cursor: not-allowed;
      background: ${disabled.background};
      ${bordersTemplate(disabled.border)};
      ${typographyTemplate(disabled.typography)};

      @media (hover: hover) and (pointer: fine) {
        &:hover {
          background: ${disabled.background};
        }
      }
    }
  `
})

export type ButtonProps = {
  /**  Specifies color */
  color?: Colors
  /** Specifies which variant to use */
  variant?: Variants
  /**
   * URL link destination
   * If defined, an 'a' element is used as root instead of 'button'
   */
  href?: string
  /** Is the button disabled */
  disabled?: boolean
  /** Type of button
   * @default 'button'
   */
  type?: string
  /** FullWidth (stretched) button  */
  fullWidth?: boolean
} & ButtonHTMLAttributes<HTMLButtonElement>

export const Button: OverridableComponent<ButtonProps, HTMLButtonElement> =
  forwardRef(function Button(
    {
      color = 'primary',
      variant = 'contained',
      children,
      disabled = false,
      href,
      tabIndex = 0,
      fullWidth = false,
      ...other
    },
    ref,
  ) {
    const { density } = useEds()
    const token = useToken({ density }, getToken(variant, color))

    const as = href && !disabled ? 'a' : other.as ? other.as : 'button'

    const type = href || other.as ? undefined : 'button'

    tabIndex = disabled ? -1 : tabIndex

    const buttonProps = {
      ref,
      as,
      href,
      type,
      disabled,
      tabIndex,
      ...other,
    }

    return (
      <ThemeProvider theme={token}>
        <ButtonBase {...buttonProps}>
          {fullWidth ? (
            <InnerFullWidth>{children}</InnerFullWidth>
          ) : (
            <Inner>{children}</Inner>
          )}
        </ButtonBase>
      </ThemeProvider>
    )
  })
