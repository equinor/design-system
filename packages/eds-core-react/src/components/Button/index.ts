import { Button as ButtonWrapper, ButtonProps } from './Button'
import { ButtonGroup, ButtonGroupProps } from './ButtonGroup/ButtonGroup'
import { ToggleButton, ToggleButtonProps } from './ToggleButton/ToggleButton'

type ButtonCompoundProps = typeof ButtonWrapper & {
  Group: typeof ButtonGroup
  Toggle: typeof ToggleButton
}

const Button = ButtonWrapper as ButtonCompoundProps
Button.Group = ButtonGroup
Button.Toggle = ToggleButton

Button.Group.displayName = 'Button.Group'
Button.Toggle.displayName = 'Button.Toggle'

export { Button }
export type { ButtonProps, ButtonGroupProps, ToggleButtonProps }
