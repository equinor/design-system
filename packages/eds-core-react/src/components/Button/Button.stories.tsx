import { useState, useEffect } from 'react'
import {
  Button,
  Icon,
  ButtonProps,
  EdsProvider,
  Density,
  Progress,
} from '../..'
import { Story, ComponentMeta } from '@storybook/react'
import { menu, add, save } from '@equinor/eds-icons'
import { Stack } from './../../../.storybook/components'
// import { Group } from '../Group'
import page from './Button.docs.mdx'

export default {
  title: 'Inputs/Button',
  component: Button,
  args: {
    as: 'button',
  },
  argTypes: {
    as: {
      options: ['span', 'a', 'button'],
      control: {
        type: 'select',
      },
    },
  },
  parameters: {
    docs: {
      page,
      source: {
        excludeDecorators: true,
      },
    },
  },
} as ComponentMeta<typeof Button>

export const Introduction: Story<ButtonProps> = (args) => {
  return <Button {...args}>You can control me</Button>
}
Introduction.args = {
  as: undefined,
}
Introduction.decorators = [
  (Story) => (
    <Stack>
      <Story />
    </Stack>
  ),
]

export const Basic: Story<ButtonProps> = () => (
  <>
    <Button>Contained</Button>
    <Button variant="contained_icon" aria-label="add action">
      <Icon data={add}></Icon>
    </Button>
    <Button variant="outlined">Outlined</Button>
    <Button variant="ghost">Ghost</Button>
    <Button variant="ghost_icon" aria-label="save action">
      <Icon data={save}></Icon>
    </Button>
  </>
)
Basic.decorators = [
  (Story) => (
    <Stack>
      <Story />
    </Stack>
  ),
]

export const IconButton: Story<ButtonProps> = () => (
  <>
    <Button variant="ghost_icon" aria-label="save action">
      <Icon data={save}></Icon>
    </Button>
    <Button variant="ghost_icon" color="secondary" aria-label="save action">
      <Icon data={save}></Icon>
    </Button>
    <Button variant="ghost_icon" color="danger" aria-label="save action">
      <Icon data={save}></Icon>
    </Button>
    <Button variant="ghost_icon" disabled aria-label="save action">
      <Icon data={save}></Icon>
    </Button>
    <Button variant="contained_icon" aria-label="add action">
      <Icon data={add}></Icon>
    </Button>
    <Button variant="contained_icon" color="secondary" aria-label="add action">
      <Icon data={add}></Icon>
    </Button>
    <Button variant="contained_icon" color="danger" aria-label="add action">
      <Icon data={add}></Icon>
    </Button>
    <Button variant="contained_icon" disabled aria-label="add action">
      <Icon data={add}></Icon>
    </Button>
  </>
)

IconButton.storyName = 'Icon button'
IconButton.decorators = [
  (Story) => (
    <Stack>
      <Story />
    </Stack>
  ),
]

export const Color: Story<ButtonProps> = () => (
  <>
    <Button color="primary">Primary</Button>
    <Button color="secondary">Secondary</Button>
    <Button color="danger">Danger</Button>
  </>
)
Color.decorators = [
  (Story) => (
    <Stack>
      <Story />
    </Stack>
  ),
]

export const Hierarchy: Story<ButtonProps> = () => (
  <>
    <Button>Contained</Button>
    <Button variant="outlined">Outlined</Button>
    <Button variant="ghost">Ghost</Button>
  </>
)
Hierarchy.decorators = [
  (Story) => (
    <Stack>
      <Story />
    </Stack>
  ),
]

export const FileUpload: Story<ButtonProps> = () => (
  <label htmlFor="file-upload">
    <input type="file" id="file-upload" style={{ display: 'none' }} multiple />
    <Button as="span">Upload file</Button>
  </label>
)
FileUpload.storyName = 'File upload'
FileUpload.decorators = [
  (Story) => (
    <Stack>
      <Story />
    </Stack>
  ),
]

export const ProgressButton: Story<ButtonProps> = () => (
  <>
    <Button>
      <Progress.Dots />
    </Button>
    <Button variant="ghost_icon">
      <Progress.Dots color="primary" />
    </Button>
    <Button>
      <Progress.Circular size={16} color="neutral" />
    </Button>
    <Button variant="ghost_icon">
      <Progress.Circular size={16} color="primary" />
    </Button>
  </>
)
ProgressButton.storyName = 'Progress button'
ProgressButton.decorators = [
  (Story) => (
    <Stack>
      <Story />
    </Stack>
  ),
]

export const All: Story<ButtonProps> = () => (
  <>
    <Button>Primary</Button>
    <Button color="secondary">Secondary</Button>
    <Button color="danger">Danger</Button>
    <Button disabled>Disabled</Button>
    <Button variant="outlined">Primary</Button>
    <Button variant="outlined" color="secondary">
      Secondary
    </Button>
    <Button variant="outlined" color="danger">
      Danger
    </Button>
    <Button variant="outlined" disabled>
      Disabled
    </Button>
    <Button variant="ghost">Primary</Button>
    <Button variant="ghost" color="secondary">
      Secondary
    </Button>
    <Button variant="ghost" color="danger">
      Danger
    </Button>
    <Button variant="ghost" disabled>
      Disabled
    </Button>
    <Button variant="ghost_icon" aria-label="save action">
      <Icon name="save" title="save action"></Icon>
    </Button>
    <Button variant="ghost_icon" color="secondary" aria-label="save action">
      <Icon data={save}></Icon>
    </Button>
    <Button variant="ghost_icon" color="danger" aria-label="save action">
      <Icon data={save}></Icon>
    </Button>
    <Button variant="ghost_icon" disabled aria-label="save action">
      <Icon data={save}></Icon>
    </Button>
    <Button variant="contained_icon" aria-label="add action">
      <Icon data={add}></Icon>
    </Button>
    <Button variant="contained_icon" color="secondary" aria-label="add action">
      <Icon data={add}></Icon>
    </Button>
    <Button variant="contained_icon" color="danger" aria-label="add action">
      <Icon data={add}></Icon>
    </Button>
    <Button variant="contained_icon" disabled aria-label="add action">
      <Icon data={add}></Icon>
    </Button>
  </>
)
All.decorators = [
  (Story) => (
    <Stack
      style={{
        display: 'grid',
        gridTemplateColumns: 'repeat(4, fit-content(100%))',
      }}
    >
      <Story />
    </Stack>
  ),
]

export const FullWidth: Story<ButtonProps> = () => (
  <>
    <Button fullWidth>Primary</Button>
    <Button color="secondary" fullWidth>
      Secondary
    </Button>
    <Button color="danger" fullWidth>
      Danger
    </Button>
    <Button disabled fullWidth>
      Disabled
    </Button>
    <Button fullWidth>
      <Icon data={save}></Icon>Primary
    </Button>
    <Button color="secondary" fullWidth>
      <Icon data={save}></Icon>Secondary
    </Button>
    <Button color="danger" fullWidth>
      <Icon data={save}></Icon>Danger
    </Button>
    <Button disabled fullWidth>
      <Icon data={save}></Icon>Disabled
    </Button>
    <Button fullWidth>
      Primary <Icon data={save}></Icon>
    </Button>
    <Button color="secondary" fullWidth>
      Secondary
      <Icon data={save}></Icon>
    </Button>
    <Button color="danger" fullWidth>
      Danger
      <Icon data={save}></Icon>
    </Button>
    <Button disabled fullWidth>
      Disabled
      <Icon data={save}></Icon>
    </Button>
  </>
)
FullWidth.storyName = 'Full width'
FullWidth.decorators = [
  (Story) => (
    <div style={{ margin: '32px', display: 'grid', gridGap: '16px' }}>
      <Story />
    </div>
  ),
]

export const Compact: Story<ButtonProps> = () => {
  const [density, setDensity] = useState<Density>('comfortable')

  useEffect(() => {
    // Simulate user change
    setDensity('compact')
  }, [density])

  return (
    <EdsProvider density={density}>
      <Button>Contained</Button>
      <Button variant="outlined">Outlined</Button>
      <Button variant="ghost">Ghost</Button>
      <Button variant="ghost_icon">
        <Icon data={menu} title="Ghost icon menu"></Icon>
      </Button>
    </EdsProvider>
  )
}
Compact.decorators = [
  (Story) => (
    <Stack>
      <Story />
    </Stack>
  ),
]

// export const ButtonGroup: Story<ButtonProps> = () => (
//   <Stack style={{ gridGap: 0 }}>
//     <Group>
//       <Button>Contained</Button>
//       <Button variant="outlined">Outlined</Button>
//       <Button variant="outlined">Outlined</Button>
//       <Button variant="outlined">Outlined</Button>
//     </Group>
//   </Stack>
// )
