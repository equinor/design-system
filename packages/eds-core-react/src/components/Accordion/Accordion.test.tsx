/* eslint-disable react/require-default-props */
import { render, screen, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'
import 'jest-styled-components'
import { axe } from 'jest-axe'
import { attach_file, notifications } from '@equinor/eds-icons'
import { Accordion } from '.'
import { Icon } from '../Icon'
import { Button } from '../Button'
import type { AccordionProps } from './Accordion.types'
import { useState } from 'react'

Icon.add({ attach_file, notifications })

beforeEach(() => {
  jest.spyOn(Math, 'random').mockReturnValue(0.12345)
})

afterEach(() => {
  jest.spyOn(Math, 'random').mockRestore()
})

type TestProps = {
  isExpanded?: boolean
} & AccordionProps

const SimpleAccordion = ({
  isExpanded = true,
  headerLevel = 'h2',
  chevronPosition = 'left',
}: TestProps) => (
  <Accordion headerLevel={headerLevel} chevronPosition={chevronPosition}>
    <Accordion.Item isExpanded={isExpanded}>
      <Accordion.Header data-testid="header1">Summary 1</Accordion.Header>
      <Accordion.Panel>Details 1</Accordion.Panel>
    </Accordion.Item>
  </Accordion>
)

const AccordionWithIcons = () => (
  <Accordion>
    <Accordion.Item>
      <Accordion.Header data-testid="header1" isExpanded>
        <Accordion.HeaderTitle>Summary</Accordion.HeaderTitle>
        <Icon name="attach_file" title="Attach file" size={16} />
        <Icon name="notifications" title="Notifications" size={16} />
      </Accordion.Header>
      <Accordion.Panel>Details</Accordion.Panel>
    </Accordion.Item>
  </Accordion>
)

describe('Accordion', () => {
  it('Matches snapshot', () => {
    const { asFragment } = render(<SimpleAccordion />)
    expect(asFragment()).toMatchSnapshot()
  })
  it('Should pass a11y test', async () => {
    const { container } = render(<SimpleAccordion />)
    expect(await axe(container)).toHaveNoViolations()
  })
  it('Should pass a11y test with icons ', async () => {
    const { container } = render(<AccordionWithIcons />)
    expect(await axe(container)).toHaveNoViolations()
  })
  /* failing test */
  /* it('Should pass a11y test with buttons ', async () => {
    const { container } = render(<AccordionWithButtons />)
    expect(await axe(container)).toHaveNoViolations()
  }) */
  it('Expands items based on prop', () => {
    render(<SimpleAccordion isExpanded />)
    const header1 = screen.getByTestId('header1')
    expect(header1).toHaveAttribute('aria-expanded', 'true')
  })
  it('Expands items on click', () => {
    render(<SimpleAccordion isExpanded={false} />)
    const header = screen.getByTestId('header1')
    fireEvent.click(header)
    expect(header).toHaveAttribute('aria-expanded', 'true')
  })
  it('triggers onToggle callback', () => {
    const mockOnToggle = jest.fn()
    render(
      <Accordion>
        <Accordion.Item isExpanded>
          <Accordion.Header onToggle={mockOnToggle} data-testid="header">
            Summary 1
          </Accordion.Header>
        </Accordion.Item>
      </Accordion>,
    )
    const header = screen.getByTestId('header')
    fireEvent.click(header)
    expect(mockOnToggle).toHaveBeenCalled()
  })

  type ControlledProps = {
    onExpandedChange: () => void
  }
  const ControlledAccordion = ({ onExpandedChange }: ControlledProps) => {
    const [expanded, setExpanded] = useState(false)
    const toggleAccordion = (state: boolean) => {
      setExpanded(state)
      onExpandedChange()
    }
    return (
      <>
        <Button data-testid="button" onClick={() => toggleAccordion(!expanded)}>
          testbutton
        </Button>
        <Accordion>
          <Accordion.Item
            isExpanded={expanded}
            onExpandedChange={toggleAccordion}
          >
            <Accordion.Header data-testid="header">Summary 1</Accordion.Header>
          </Accordion.Item>
        </Accordion>
      </>
    )
  }

  it('Can be a controlled component', () => {
    const mockToggled = jest.fn()
    render(<ControlledAccordion onExpandedChange={mockToggled} />)
    const header = screen.getByTestId('header')
    const button = screen.getByTestId('button')
    expect(header).not.toHaveAttribute('aria-expanded', 'true')
    fireEvent.click(button)
    expect(mockToggled).toHaveBeenCalled()
    expect(header).toHaveAttribute('aria-expanded', 'true')
  })
  it('Add custom icons', () => {
    render(<AccordionWithIcons />)
    const header = screen.getByTestId('header1')
    const icon = screen.getAllByTitle('Notifications')
    expect(header).toContainElement(icon[0])
  })
  // it('Add custom buttons', () => {
  //   render(<AccordionWithButtons />)
  //   expect(screen.getAllByRole('button')).toHaveLength(3) // 2 + itself
  // })
  // it('Does not expand when clicking custom buttons', () => {
  //   render(<AccordionWithButtons />)
  //   const header = screen.getByTestId('header1')
  //   const button = screen.getByTestId('button2')
  //   fireEvent.click(button)
  //   expect(header).toHaveAttribute('aria-expanded', 'false')
  // })
})
