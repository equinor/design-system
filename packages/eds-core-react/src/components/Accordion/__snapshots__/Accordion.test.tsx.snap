// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion Matches snapshot 1`] = `
<DocumentFragment>
  .c2 {
  transform: none;
}

.c4 {
  display: -webkit-box;
  -webkit-line-clamp: 1;
  -webkit-box-orient: vertical;
  flex-grow: 1;
  overflow: hidden;
  text-align: left;
  color: var(--eds_interactive_primary__resting, rgba(0, 112, 121, 1));
}

.c0 {
  margin: 0;
  padding: 0;
  height: 48px;
  box-sizing: border-box;
  display: flex;
  flex-wrap: nowrap;
  justify-content: space-between;
  background-color: var(--eds_ui_background__default, rgba(255, 255, 255, 1));
  border: 1px solid var(--eds_ui_background__medium, rgba(220, 220, 220, 1));
  color: var(--eds_ui__accordion_header_color, rgba(61, 61, 61, 1));
  cursor: pointer;
}

.c1 {
  margin: 0;
  color: var(--eds_ui__accordion_header_color, rgba(61, 61, 61, 1));
  font-family: Equinor;
  font-size: 1.000rem;
  font-weight: 500;
  line-height: 1.500em;
  text-align: left;
  padding-left: 16px;
  padding-right: 16px;
  border: 0;
  background-color: transparent;
  margin: 0;
  display: flex;
  align-items: center;
  flex-grow: 1;
  color: var(--eds_ui__accordion_header_color, rgba(61, 61, 61, 1));
  cursor: pointer;
}

.c1[data-focus-visible-added]:focus {
  outline: 2px dashed var(--eds_interactive__focus, rgba(0, 112, 121, 1));
  outline-offset: 2px;
}

.c1:focus-visible {
  outline: 2px dashed var(--eds_interactive__focus, rgba(0, 112, 121, 1));
  outline-offset: 2px;
}

.c1 >svg {
  color: var(--eds_text__static_icons__default, rgba(61, 61, 61, 1));
}

.c3 {
  flex-shrink: 0;
  margin-right: 32px;
}

.c5 {
  border: 1px solid var(--eds_ui_background__medium, rgba(220, 220, 220, 1));
  border-top: 0;
  padding-left: 16px;
  padding-top: 16px;
  padding-right: 16px;
  padding-bottom: 16px;
  margin: 0;
  color: var(--eds_input__text_color, rgba(0, 0, 0, 1));
  font-family: Equinor;
  font-size: 1.000rem;
  font-weight: 400;
  line-height: 1.500em;
  letter-spacing: 0.025em;
  text-align: left;
  background: var(--eds_ui_background__default, rgba(255, 255, 255, 1));
  min-height: 96px;
  box-sizing: border-box;
}

@media (hover:hover) and (pointer:fine) {
  .c0:hover {
    background: var(--eds_ui_background__light, rgba(247, 247, 247, 1));
  }
}

<div>
    <div>
      <h2
        class="c0"
      >
        <button
          aria-controls="eds-accordion-12345-panel-1"
          aria-expanded="true"
          class="c1"
          data-testid="header1"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="c2 c3"
            fill="var(--eds_interactive_primary__resting, rgba(0, 112, 121, 1))"
            height="24px"
            viewBox="0 0 24 24"
            width="24px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              class=""
              clip-rule="evenodd"
              d="m7.41 15.705 4.59-4.58 4.59 4.58 1.41-1.41-6-6-6 6 1.41 1.41Z"
              data-testid="eds-icon-path"
              fill-rule="evenodd"
            />
          </svg>
          <span
            class="c4"
          >
            Summary 1
          </span>
        </button>
      </h2>
      <div
        aria-labelledby="eds-accordion-12345-header-1"
        class="c5"
        id="eds-accordion-12345-panel-1"
        role="region"
      >
        Details 1
      </div>
    </div>
  </div>
</DocumentFragment>
`;
